### ISSUE TITLE
[CCI超买超卖策略] | CCI Overbought Oversold

### ISSUE THREAD

**Issue by Kaoruha** [Kaoruha/GinkGO#4466]
**标签**: quantitative-strategy
### 摘要
基于商品通道指数(CCI)的经典超买超卖信号，结合自适应参数和趋势过滤构建趋势跟踪策略。
### 算法描述
**核心原理**：
CCI衡量价格相对于其平均统计区间偏离程度，当CCI突破±100阈值时表明趋势强度异常。

**CCI计算公式**：
CCI = (Typical_Price - SMA_20) / (0.015 * Mean_Deviation)

其中：
- Typical_Price = (High + Low + Close) / 3
- 0.015 确保80%的值落在±100内

**改进的CCI信号**：
1. **经典阈值法**：
   - CCI > +100：超买 → 做空信号
   - CCI < -100：超卖 → 做多信号
2. **零轴穿越法**：
   - CCI上穿0 → 动量转多
   - CCI下穿0 → 动量转空
3. **背离检测**：
   - 价格创新高但CCI下降 → 顶背离
   - 价格创新低但CCI上升 → 底背离

**自适应性**：
- 波动率高时扩大窗口
- 趋势明确时收紧阈值

**风控要点**：
- 入场后CCI回撤超过50%止损
- 浮动止损跟踪趋势
- 多市场相关性过滤
### Python代码示例
```python
import numpy as np
import pandas as pd
from datetime import datetime, timedelta
from typing import Dict, List, Tuple, Optional
from dataclasses import dataclass
from enum import Enum

@dataclass
class CCIResult:
    timestamp: datetime
    cci: float
    sma: float
    mean_dev: float
    signal: int  # 1=long, -1=short, 0=neutral
    strength: float  # 0-1

@dataclass
class TradingSignal:
    timestamp: datetime
    direction: int
    entry_price: float
    stop_loss: float
    take_profit: float
    confidence: float
    indicators: Dict

class CommodityChannelIndex:
    """商品通道指数计算器"""
    
    def __init__(self, period: int = 20):
        self.period = period
    
    @staticmethod
    def typical_price(high: float, low: float, close: float) -> float:
        """典型价格"""
        return (high + low + close) / 3.0
    
    def calculate_sma(self, prices: np.ndarray) -> float:
        """简单移动平均"""
        return np.mean(prices[-self.period:])
    
    def calculate_mean_deviation(self, prices: np.ndarray, sma: float) -> float:
        """平均偏差"""
        deviations = np.abs(prices[-self.period:] - sma)
        return np.mean(deviations)
    
    def calculate_cci(
        self,
        high: float,
        low: float,
        close: float,
        sma: float,
        mean_dev: float
    ) -> float:
        """计算CCI"""
        tp = self.typical_price(high, low, close)
        
        if abs(mean_dev) < 1e-10:
            return 0.0
        
        cci = (tp - sma) / (0.015 * mean_dev)
        return cci
    
    def calculate(
        self,
        data: pd.DataFrame
    ) -> List[CCIResult]:
        """批量计算CCI"""
        results = []
        
        typical_prices = (data["high"] + data["low"] + data["close"]) / 3.0
        
        for i in range(len(data)):
            if i < self.period - 1:
                continue
            
            window = typical_prices[max(0, i-self.period+1):i+1].values
            sma = self.calculate_sma(window)
            mean_dev = self.calculate_mean_deviation(window, sma)
            
            cci = self.calculate_cci(
                data.iloc[i]["high"],
                data.iloc[i]["low"],
                data.iloc[i]["close"],
                sma,
                mean_dev
            )
            
            # 生成信号
            signal = 0
            if cci > 100:
                signal = -1  # 超买，做空
            elif cci < -100:
                signal = 1   # 超卖，做多
            
            strength = min(1.0, abs(cci - 100) / 100) if abs(cci) > 100 else 0.0
            
            results.append(CCIResult(
                timestamp=data.iloc[i]["timestamp"],
                cci=cci,
                sma=sma,
                mean_dev=mean_dev,
                signal=signal,
                strength=strength
            ))
        
        return results

class AdaptiveCCI:
    """自适应CCI策略"""
    
    def __init__(
        self,
        base_period: int = 20,
        min_period: int = 10,
        max_period: int = 40,
        upper_threshold: float = 100.0,
        lower_threshold: float = -100.0,
        atr_period: int = 14,
        atr_multiplier: float = 2.0
    ):
        self.base_period = base_period
        self.min_period = min_period
        self.max_period = max_period
        self.upper_threshold = upper_threshold
        self.lower_threshold = lower_threshold
        self.atr_period = atr_period
        self.atr_multiplier = atr_multiplier
        
        self.cci_calculator = CommodityChannelIndex(base_period)
        self.price_history: List[float] = []
        self.atr_history: List[float] = []
    
    def calculate_atr(
        self,
        highs: np.ndarray,
        lows: np.ndarray,
        closes: np.ndarray
    ) -> float:
        """计算ATR"""
        if len(closes) < self.atr_period + 1:
            return 0.0
        
        trs = []
        for i in range(1, len(closes)):
            high = highs[i]
            low = lows[i]
            prev_close = closes[i-1]
            
            tr = max(
                high - low,
                abs(high - prev_close),
                abs(low - prev_close)
            )
            trs.append(tr)
        
        return np.mean(trs[-self.atr_period:])
    
    def adaptive_period(self, volatility: float, base_volatility: float) -> int:
        """根据波动率自适应调整周期"""
        if base_volatility < 1e-10:
            return self.base_period
        
        ratio = volatility / base_volatility
        
        if ratio > 1.5:
            return min(self.max_period, int(self.base_period * ratio))
        elif ratio < 0.7:
            return max(self.min_period, int(self.base_period * ratio))
        else:
            return self.base_period
    
    def detect_divergence(
        self,
        prices: np.ndarray,
        ccis: np.ndarray,
        window: int = 10
    ) -> Tuple[str, float]:
        """检测背离"""
        if len(prices) < window * 2 or len(ccis) < window * 2:
            return "none", 0.0
        
        recent_prices = prices[-window:]
        recent_cci = ccis[-window:]
        
        # 价格创新高
        price_high_idx = np.argmax(recent_prices)
        
        # CCI对应位置
        cci_corresponding = recent_cci[price_high_idx]
        cci_later = recent_cci[-1]
        
        # 顶背离
        if price_high_idx == len(recent_prices) - 1:
            if cci_later < cci_corresponding:
                return "bearish", 0.7
        
        # 价格创新低
        price_low_idx = np.argmin(recent_prices)
        cci_corresponding_low = recent_cci[price_low_idx]
        cci_later_low = recent_cci[-1]
        
        # 底背离
        if price_low_idx == len(recent_prices) - 1:
            if cci_later_low > cci_corresponding_low:
                return "bullish", 0.7
        
        return "none", 0.0
    
    def generate_signal(
        self,
        data: pd.DataFrame,
        trend_filter: bool = True
    ) -> List[TradingSignal]:
        """生成交易信号"""
        signals = []
        
        cci_results = self.cci_calculator.calculate(data)
        
        highs = data["high"].values
        lows = data["low"].values
        closes = data["close"].values
        
        # 计算ATR
        atr = self.calculate_atr(highs, lows, closes)
        self.atr_history.append(atr)
        
        base_atr = np.mean(self.atr_history[-20:]) if len(self.atr_history) >= 20 else atr
        
        for i, cci_result in enumerate(cci_results):
            if i < 2:
                continue
            
            prev_cci = cci_results[i-1].cci
            curr_cci = cci_result.cci
            
            # 获取价格数据
            entry_price = closes[i]
            high_price = highs[i]
            low_price = lows[i]
            
            # 趋势过滤（简单移动平均线）
            sma_50 = np.mean(closes[max(0, i-50):i]) if i >= 50 else entry_price
            trend_up = entry_price > sma_50
            
            # 经典CCI信号
            signal = 0
            if prev_cci < self.upper_threshold and curr_cci >= self.upper_threshold:
                signal = -1  # 上穿+100，做空
            elif prev_cci > self.lower_threshold and curr_cci <= self.lower_threshold:
                signal = 1   # 下穿-100，做多
            
            # 趋势过滤
            if trend_filter:
                if signal == 1 and not trend_up:
                    signal = 0  # 排除下跌趋势中的做多
                elif signal == -1 and trend_up:
                    signal = 0  # 排除上涨趋势中的做空
            
            # 背离检测
            divergence_type, div_strength = self.detect_divergence(
                closes[:i+1],
                np.array([r.cci for r in cci_results[:i+1]]),
                window=10
            )
            
            if divergence_type == "bearish":
                signal = -1
            elif divergence_type == "bullish":
                signal = 1
            
            # 设置止损止盈
            stop_loss = 0.0
            take_profit = 0.0
            
            if signal == 1:  # 做多
                stop_loss = entry_price - atr * self.atr_multiplier
                take_profit = entry_price + atr * self.atr_multiplier * 2
            elif signal == -1:  # 做空
                stop_loss = entry_price + atr * self.atr_multiplier
                take_profit = entry_price - atr * self.atr_multiplier * 2
            
            # 置信度
            confidence = cci_result.strength
            if divergence_type != "none":
                confidence += div_strength
            confidence = min(1.0, confidence)
            
            signals.append(TradingSignal(
                timestamp=cci_result.timestamp,
                direction=signal,
                entry_price=entry_price,
                stop_loss=stop_loss,
                take_profit=take_profit,
                confidence=confidence,
                indicators={
                    "cci": curr_cci,
                    "divergence": divergence_type,
                    "atr": atr
                }
            ))
        
        return signals
    
    def backtest(
        self,
        signals: List[TradingSignal],
        initial_capital: float = 100000,
        commission: float = 0.001
    ) -> Dict:
        """回测"""
        capital = initial_capital
        position = None
        trades = []
        equity_curve = []
        
        for signal in signals:
            if signal.direction == 0:
                equity_curve.append(capital)
                continue
            
            if position is None:
                if signal.direction == 1:
                    position = {
                        "type": "long",
                        "entry_price": signal.entry_price,
                        "entry_time": signal.timestamp,
                        "stop_loss": signal.stop_loss,
                        "take_profit": signal.take_profit,
                        "shares": int(capital * 0.1 / signal.entry_price)
                    }
                elif signal.direction == -1:
                    position = {
                        "type": "short",
                        "entry_price": signal.entry_price,
                        "entry_time": signal.timestamp,
                        "stop_loss": signal.stop_loss,
                        "take_profit": signal.take_profit,
                        "shares": int(capital * 0.1 / signal.entry_price)
                    }
            else:
                # 检查止损止盈
                should_close = False
                pnl = 0
                
                if position["type"] == "long":
                    if signal.entry_price <= position["stop_loss"]:
                        pnl = (position["stop_loss"] - position["entry_price"]) * position["shares"]
                        should_close = True
                    elif signal.entry_price >= position["take_profit"]:
                        pnl = (position["take_profit"] - position["entry_price"]) * position["shares"]
                        should_close = True
                else:  # short
                    if signal.entry_price >= position["stop_loss"]:
                        pnl = (position["entry_price"] - position["stop_loss"]) * position["shares"]
                        should_close = True
                    elif signal.entry_price <= position["take_profit"]:
                        pnl = (position["entry_price"] - position["take_profit"]) * position["shares"]
                        should_close = True
                
                if should_close:
                    capital += pnl - commission * capital
                    trades.append({
                        "entry_time": position["entry_time"],
                        "exit_time": signal.timestamp,
                        "pnl": pnl,
                        "type": position["type"]
                    })
                    position = None
            
            equity_curve.append(capital)
        
        return {
            "final_capital": capital,
            "total_return": (capital - initial_capital) / initial_capital,
            "num_trades": len(trades),
            "win_rate": sum(1 for t in trades if t["pnl"] > 0) / max(1, len(trades)),
            "equity_curve": equity_curve
        }

# 测试
if __name__ == "__main__":
    np.random.seed(42)
    
    dates = pd.date_range("2024-01-01", "2025-01-01", freq="D")
    n = len(dates)
    
    # 生成趋势性价格数据
    trend = np.linspace(0, 10, n)
    cycles = 5 * np.sin(np.linspace(0, 4*np.pi, n))
    noise = np.random.randn(n) * 2
    
    close_prices = 100 + trend + cycles + noise
    
    # 生成高低收
    high_prices = close_prices + np.abs(np.random.rand(n) * 3)
    low_prices = close_prices - np.abs(np.random.rand(n) * 3)
    
    data = pd.DataFrame({
        "timestamp": dates,
        "high": high_prices,
        "low": low_prices,
        "close": close_prices
    })
    
    strategy = AdaptiveCCI()
    signals = strategy.generate_signal(data)
    
    result = strategy.backtest(signals)
    
    print(f"Final Capital: ${result['final_capital']:,.2f}")
    print(f"Total Return: {result['total_return']:.2%}")
    print(f"Win Rate: {result['win_rate']:.2%}")
